monai/apps/generation/maisi/networks/autoencoderkl_maisi.py

```diff
@@ -0,0 +1,901 @@
+
+        return input
+
+    def _cat_inputs(self, inputs):
+        input_type = inputs[0].device.type
+        input = inputs[0].clone().to("cpu", non_blocking=True) if input_type == "cuda" else inputs[0].clone()
+        inputs[0] = 0
+        _empty_cuda_cache()
+
+        for k in range(len(inputs) - 1):
+            input = torch.cat((input, inputs[k + 1].cpu()), dim=1)
+            inputs[k + 1] = 0
+            _empty_cuda_cache()
+            gc.collect()
+
+            if self.debug:
+                print(f"MaisiGroupNorm3D cat: {k + 1}/{len(inputs) - 1}.")
+
+        return input.to("cuda", non_blocking=True) if input_type == "cuda" else input
+
+
+class MaisiConvolution(nn.Module):
+    """
+    Convolutional layer with optional debug output and custom splitting mechanism.
+
+    Args:
+        spatial_dims: Number of spatial dimensions (1D, 2D, 3D).
+        in_channels: Number of input channels.
+        out_channels: Number of output channels.
+        num_splits: Number of splits for the input tensor.
+        dim_split: Dimension of splitting for the input tensor.
+        debug: Whether to print debug information.
+        Additional arguments for the convolution operation.
+        https://docs.monai.io/en/stable/networks.html#convolution
+    """
+
+    def __init__(
+        self,
+        spatial_dims: int,
+        in_channels: int,
+        out_channels: int,
+        num_splits: int,
+        dim_split: int,
+        debug: bool,
+        strides: Sequence[int] | int = 1,
+        kernel_size: Sequence[int] | int = 3,
+        adn_ordering: str = "NDA",
+        act: tuple | str | None = "PRELU",
+        norm: tuple | str | None = "INSTANCE",
+        dropout: tuple | str | float | None = None,
+        dropout_dim: int = 1,
+        dilation: Sequence[int] | int = 1,
+        groups: int = 1,
+        bias: bool = True,
+        conv_only: bool = False,
+        is_transposed: bool = False,
+        padding: Sequence[int] | int | None = None,
+        output_padding: Sequence[int] | int | None = None,
+    ) -> None:
+        super().__init__()
+        self.conv = Convolution(
+            spatial_dims=spatial_dims,
+            in_channels=in_channels,
+            out_channels=out_channels,
+            strides=strides,
+            kernel_size=kernel_size,
+            adn_ordering=adn_ordering,
+            act=act,
+            norm=norm,
+            dropout=dropout,
+            dropout_dim=dropout_dim,
+            dilation=dilation,
+            groups=groups,
+            bias=bias,
+            conv_only=conv_only,
+            is_transposed=is_transposed,
+            padding=padding,
+            output_padding=output_padding,
+        )
+
+        self.dim_split = dim_split
+        self.stride = strides[self.dim_split] if isinstance(strides, list) else strides
+        self.num_splits = num_splits
+        self.debug = debug
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
```

Comment:
Hi @dongyang0122 , the `forward` logic seems complicated at the first sight. Maybe we can group the logics into smaller functions? For example:

```python
def forward(self, x: torch.Tensor) -> torch.Tensor:
    num_splits = self.num_splits
    debug_print(self.debug, f"num_splits: {num_splits}")

    split_size, padding = calculate_split_size_and_padding(x, self.dim_split, num_splits, self.stride)
    debug_print(self.debug, f"padding: {padding}")

    splits = split_tensor(x, num_splits, split_size, padding, self.dim_split)
    debug_print(self.debug, "splits", splits)

    del x
    _empty_cuda_cache()

    outputs = [self.conv(split) for split in splits]
    debug_print(self.debug, "outputs before", outputs)

    split_size_out, padding_s = adjust_output_dimensions(outputs, splits, self.dim_split)
    outputs = trim_and_prepare_outputs(outputs, num_splits, split_size_out, padding_s, self.dim_split)
    debug_print(self.debug, "outputs after", outputs)

    x = concatenate_outputs(outputs, self.dim_split)
    del outputs
    _empty_cuda_cache()

    return x
```

Reply:
Hi @mingxin-zheng, I created two functions about splitting and concatenating, and added necessary comments. Hopefully it would make the script less complicated. Please let me know if you have further suggestions. Thanks.

Reply:
Thanks @dongyang0122 . Looks good to me on the smaller functions created.